language/move-binary-format/src/check_bounds.rs

```diff
@@ -19,100 +20,158 @@ use move_core_types::vm_status::StatusCode;
     current_function: Option<FunctionDefinitionIndex>,
 }
 
 impl<'a> BoundsChecker<'a> {
-    pub fn verify(module: &'a CompiledModuleMut) -> PartialVMResult<()> {
+    pub fn verify_script(script: &'a CompiledScript) -> PartialVMResult<()> {
         let mut bounds_check = Self {
-            module,
+            view: BinaryIndexedView::Script(&script),
             current_function: None,
         };
         // TODO: this will not be true once we change CompiledScript and remove the
         // FunctionDefinition for `main`
-        if bounds_check.module.module_handles.is_empty() {
+        if bounds_check.view.module_handles().is_empty() {
             let status =
                 verification_error(StatusCode::NO_MODULE_HANDLES, IndexKind::ModuleHandle, 0);
             return Err(status);
         }
 
-        for signature in &bounds_check.module.signatures {
+        for signature in bounds_check.view.signatures() {
             bounds_check.check_signature(signature)?
         }
-        for constant in &bounds_check.module.constant_pool {
+        for constant in bounds_check.view.constant_pool() {
             bounds_check.check_constant(constant)?
         }
-        for module_handle in &bounds_check.module.module_handles {
+        for script_handle in bounds_check.view.module_handles() {
+            bounds_check.check_module_handle(script_handle)?
+        }
+        for struct_handle in bounds_check.view.struct_handles() {
+            bounds_check.check_struct_handle(struct_handle)?
+        }
+        for function_handle in bounds_check.view.function_handles() {
+            bounds_check.check_function_handle(function_handle)?
+        }
+        for field_handle in bounds_check.view.field_handles().unwrap_or(&[]) {
+            bounds_check.check_field_handle(field_handle)?
+        }
+        for friend_decl in bounds_check.view.friend_decls().unwrap_or(&[]) {
+            bounds_check.check_module_handle(friend_decl)?
+        }
+        for struct_instantiation in bounds_check.view.struct_instantiations().unwrap_or(&[]) {
+            bounds_check.check_struct_instantiation(struct_instantiation)?
+        }
+        for function_instantiation in bounds_check.view.function_instantiations() {
+            bounds_check.check_function_instantiation(function_instantiation)?
+        }
+        for field_instantiation in bounds_check.view.field_instantiations().unwrap_or(&[]) {
+            bounds_check.check_field_instantiation(field_instantiation)?
+        }
+        for struct_def in bounds_check.view.struct_defs().unwrap_or(&[]) {
+            bounds_check.check_struct_def(struct_def)?
+        }
+        let view = bounds_check.view;
+        for (function_def_idx, function_def) in
+            view.function_defs().unwrap_or(&[]).iter().enumerate()
+        {
+            bounds_check.check_function_def(function_def_idx, function_def)?
+        }
+        bounds_check.check_self_module_handle()
```

Comment:
It looks like most of this code is shared between the two. Maybe just have a function for all the shared stuff and then keep the non-overlapping code in `verify_module/script`